### PR TITLE
[REVIEW] Remove making redundant `copy` across code-base

### PR DESCRIPTION
**Pull request by galipremsagar**
This PR removes some of the existing `copy` overheads in the code-base:

All scan operations with `skipna=False`:

```python
# branch-22.04:

In [3]: df = cudf.DataFrame(
   ...:     {
   ...:         "a": [1, 2, 3, np.nan] * 10000000,
   ...:         "c": [0.0, 0.12, np.nan, 10.12] * 10000000,
   ...:     },
   ...:     nan_as_null=False,
   ...: )


In [3]: %timeit df.cummin(skipna=False)
27.6 ms ± 587 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)



# THIS PR:

In [3]: %timeit df.cummin(skipna=False)
14.3 ms ± 43 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

```

`assign`:
```python
# branch-22.04:


In [3]: %timeit df.assign(b=10, c=11, d=12)
36.5 ms ± 1.16 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)


# THIS PR:

In [3]: %timeit df.assign(b=10, c=11, d=12)
27.6 ms ± 8.73 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

```

`replace`:

```python

In [10]: df
Out[10]: 
          a      c
0         1   0.00
1         2   0.12
2         3  10.12
3         1   0.00
4         2   0.12
...      ..    ...
29999995  2   0.12
29999996  3  10.12
29999997  1   0.00
29999998  2   0.12
29999999  3  10.12

[30000000 rows x 2 columns]

# branch-22.04
In [11]: %timeit df.replace([1, 2, 3], [4, 5, 6])
25.7 ms ± 157 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)


# THIS PR:

In [3]: %timeit df.replace([1, 2, 3], [4, 5, 6])
14.5 ms ± 52.1 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)


```
`nans_to_nulls`:
```python
# branch-22.04:


In [3]: df = cudf.DataFrame(
   ...:     {
   ...:         "a": [1, 2, 3, np.nan] * 10000000,
   ...:         "b": ["a", "b", "c", "d"] * 10000000,
   ...:         "c": [0.0, 0.12, np.nan, 10.12] * 10000000,
   ...:     },
   ...:     nan_as_null=False,
   ...: )


In [4]: %timeit df.nans_to_nulls()
32.1 ms ± 269 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)


# THIS PR:


In [5]: %timeit df.nans_to_nulls()
10.3 ms ± 28.8 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```

`Series.astype`:

```python
# branchh-22.04:

In [11]: s.nans_to_nulls()
Out[11]: 
0            1.0
1            2.0
2            3.0
3           <NA>
4            1.0
            ... 
39999995    <NA>
39999996     1.0
39999997     2.0
39999998     3.0
39999999    <NA>
Name: a, Length: 40000000, dtype: float64

In [12]: s = s.nans_to_nulls()

In [11]: %timeit s.astype('int64', copy=True)
18.2 ms ± 20.7 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)


# THIS PR:

In [8]: %timeit s.astype('int64', copy=True)
9.19 ms ± 21.7 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```

`Series.add_prefix/add_suffix`:

```python

In [8]: s
Out[8]: 
0           a
1           b
2           c
3           d
4           a
           ..
39999995    d
39999996    a
39999997    b
39999998    c
39999999    d
Name: b, Length: 40000000, dtype: object

In [6]: s.index
Out[6]: 
StringIndex(['0' '1' '2' ... '39999997' '39999998' '39999999'], dtype='object')


# branch-22.04:

In [4]: %timeit s.add_prefix("abcdef")
63.5 ms ± 281 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)

# THIS PR:
In [7]: %timeit s.add_prefix("abcdef")
60.9 ms ± 64.5 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
```

### CODE REVIEW

File: python/cudf/cudf/core/frame.py [hunk 1]
```diff
@@ -4826,13 +4828,16 @@ def nans_to_nulls(self):
...
-                name: col.copy().nans_to_nulls()
-                for name, col in self._data.items()
-            },
-            self._index,
-        )
+        result_data = {}
+        for name, col in self._data.items():
+            if (
+                isinstance(col, cudf.core.column.NumericalColumn)
+                and col.nan_count
```
Comment (galipremsagar) [inline python/cudf/cudf/core/frame.py:4835]: This makes me think should we have a `nan_count` property defined in `ColumnBase` to return `0`? @shwina @vyasr

Comment (shwina) [reply python/cudf/cudf/core/frame.py:4835]: I think:

- The check for  `nan_count`  should be in `NumericalColumn.nans_to_nulls()`
- Rather than the `isinstance` check, we could do a `try...except`:

```python
try:
    result_data[name] = col.nans_to_nulls()
except AttributeError:
    result_data[name] = col.copy()
else:
    result_data[name] = col.copy()
```

File: python/cudf/cudf/core/column/categorical.py [hunk 2]
```diff
@@ -1507,9 +1509,15 @@ def _set_categories(
...
-        old_df = cudf.DataFrame({"old_codes": old_codes, "cats": cur_cats})
-        cur_df = cudf.DataFrame({"old_codes": cur_codes, "order": cur_order})
+        new_df = cudf.DataFrame._from_data(
+            data={"new_codes": new_codes, "cats": new_cats}
+        )
+        old_df = cudf.DataFrame._from_data(
+            data={"old_codes": old_codes, "cats": cur_cats.copy()}
+        )
+        cur_df = cudf.DataFrame._from_data(
+            data={"old_codes": cur_codes.copy(), "order": cur_order}
```
Comment (vyasr) [inline python/cudf/cudf/core/column/categorical.py:1519]: Same as above, do you need to copy.

File: python/cudf/cudf/core/column/categorical.py [hunk 3]
```diff
@@ -1095,18 +1094,21 @@ def find_and_replace(
         # map it to the new label it is to be replaced by
         dtype_replace = cudf.Series(replacement_col)
         dtype_replace[dtype_replace.isin(old_cats["cats"])] = None
-        new_cats["cats"] = new_cats["cats"].replace(
+        old_cats._data["new_cats"] = old_cats["cats"].replace(
```
Comment (bdice) [inline python/cudf/cudf/core/column/categorical.py:1097]: It's awkward to have a column `"new_cats"` in a dataframe with the variable name `old_cats`. Would it make more sense to leave this in a variable `new_cats_col`, rather than add a column in the `old_cats` dataframe named `new_cats`? I see it is used immediately below to construct the DataFrame `new_cats`. I don't immediately see a good way to resolve all of the awkward name reuse.

File: python/cudf/cudf/core/column/column.py [hunk 4]
```diff
@@ -179,32 +179,24 @@ def equals(self, other: ColumnBase, check_dtypes: bool = False) -> bool:
     def all(self, skipna: bool = True) -> bool:
         # If all entries are null the result is True, including when the column
         # is empty.
-        result_col = self.nans_to_nulls() if skipna else self
```
Comment (vyasr) [inline python/cudf/cudf/core/column/column.py:182]: Why this change?

Comment (vyasr) [reply python/cudf/cudf/core/column/column.py:182]: I'm confused by the same below?

Comment (vyasr) [reply python/cudf/cudf/core/column/column.py:182]: Oh I see this method is empty and has just been removed as we discussed and only is defined for numerical now. OK we're good.

File: python/cudf/cudf/core/column/column.py [hunk 5]
```diff
@@ -179,32 +179,24 @@ def equals(self, other: ColumnBase, check_dtypes: bool = False) -> bool:
...
-        if isinstance(result_col, ColumnBase):
-            return libcudf.reduce.reduce("any", result_col, dtype=np.bool_)
-
-        return result_col
+        return libcudf.reduce.reduce("any", self, dtype=np.bool_)
 
     def dropna(self, drop_nan: bool = False) -> ColumnBase:
-        col = self.nans_to_nulls() if drop_nan else self
-        return drop_nulls([col])[0]
+        return drop_nulls([self])[0]
```
Comment (bdice) [inline python/cudf/cudf/core/column/column.py:199]:
```suggestion
        # The drop_nan argument is only used for numerical columns.
        return drop_nulls([self])[0]
```

File: python/cudf/cudf/core/column/column.py [hunk 6]
```diff
@@ -179,32 +179,24 @@ def equals(self, other: ColumnBase, check_dtypes: bool = False) -> bool:
     def all(self, skipna: bool = True) -> bool:
         # If all entries are null the result is True, including when the column
         # is empty.
```
Comment (bdice) [inline python/cudf/cudf/core/column/column.py:181]:
```suggestion
        # The skipna argument is only used for numerical columns.
        # If all entries are null the result is True, including when the column
        # is empty.
```

File: python/cudf/cudf/core/dataframe.py [hunk 7]
```diff
@@ -1321,10 +1321,16 @@ def assign(self, **kwargs):
...
+            if name in kwargs:
+                new_df[name] = kwargs.pop(name)
+            else:
+                new_df._data[name] = col.copy()
+
         for k, v in kwargs.items():
-            new[k] = v
-        return new
+            new_df[k] = v
+        return new_df
```
Comment (vyasr) [inline python/cudf/cudf/core/dataframe.py:1333]: Can we build up a dict instead of directly building up a DataFrame? That avoids calling the DataFrame constructor like we're doing everywhere else in this PR. We can also avoid all the expensive `__setitem__` calls.

File: python/cudf/cudf/core/column/numerical.py [hunk 8]
```diff
@@ -355,9 +381,15 @@ def _process_for_reduction(
...
-        )
+
+        if skipna:
+            return super(
+                NumericalColumn, self.nans_to_nulls()
+            )._process_for_reduction(skipna=skipna, min_count=min_count)
+        else:
+            return super()._process_for_reduction(
+                skipna=skipna, min_count=min_count
+            )
```
Comment (bdice) [inline python/cudf/cudf/core/column/numerical.py:392]:
```suggestion
        super_obj = super(NumericalColumn, self.nans_to_nulls()) if skipna else super()
        return super_obj._process_for_reduction(skipna=skipna, min_count=min_count)
```

File: python/cudf/cudf/core/column/numerical.py [hunk 9]
```diff
@@ -403,17 +435,16 @@ def find_and_replace(
...
                 )
             )
         elif len(replacement_col) == 1 and len(to_replace_col) == 0:
-            return replaced
+            return self.copy()
         to_replace_col, replacement_col, replaced = numeric_normalize_types(
-            to_replace_col, replacement_col, replaced
+            to_replace_col, replacement_col, self
         )
         df = cudf.DataFrame({"old": to_replace_col, "new": replacement_col})
```
Comment (bdice) [inline python/cudf/cudf/core/column/numerical.py:449]: Following the changes elsewhere in this PR, I think this can be updated as well:
```suggestion
        df = cudf.DataFrame._from_data(
            {"old": to_replace_col, "new": replacement_col}
        )
```

File: python/cudf/cudf/core/frame.py [hunk 10]
```diff
@@ -4838,13 +4843,13 @@ def nans_to_nulls(self):
...
-            },
-            self._index,
-        )
+        result_data = {}
+        for name, col in self._data.items():
+            try:
+                result_data[name] = col.nans_to_nulls()
+            except AttributeError:
+                result_data[name] = col.copy()
+        return self._from_data(result_data, self._index,)
```
Comment (bdice) [inline python/cudf/cudf/core/frame.py:4852]:
```suggestion
        return self._from_data(result_data, self._index)
```

File: python/cudf/cudf/core/series.py [hunk 11]
```diff
@@ -1908,10 +1908,7 @@ def astype(self, dtype, copy=False, errors="raise"):
         try:
             data = self._column.astype(dtype)
 
-            return self._from_data(
-                {self.name: (data.copy(deep=True) if copy else data)},
-                index=self._index,
-            )
+            return self._from_data({self.name: data}, index=self._index,)
```
Comment (bdice) [inline python/cudf/cudf/core/series.py:1911]:
```suggestion
            return self._from_data({self.name: data}, index=self._index)
```

**Review by bdice — COMMENTED** [review]
A few suggestions, overall this looks great. Lots of eliminated copies! 👍

File: python/cudf/cudf/core/multiindex.py [hunk 12]
```diff
@@ -1,4 +1,4 @@
-# Copyright (c) 2019-2021, NVIDIA CORPORATION.
+# Copyright (c) 2019-2022, NVIDIA CORPORATION.
```
Comment (vyasr) [inline python/cudf/cudf/core/multiindex.py:1]: ?

File: python/cudf/cudf/core/column/categorical.py (hunk 3 above)
Comment (galipremsagar) [reply python/cudf/cudf/core/column/categorical.py]: Changed it 👍

File: python/cudf/cudf/core/column/numerical.py (hunk 9 above)
Comment (galipremsagar) [reply python/cudf/cudf/core/column/numerical.py:452]: Yup, missed it

File: python/cudf/cudf/core/column/categorical.py [hunk 13]
```diff
@@ -1080,12 +1079,12 @@ def find_and_replace(
...
-        old_cats = cudf.DataFrame()
-        old_cats["cats"] = column.as_column(replaced.dtype.categories)
-        new_cats = old_cats.copy(deep=True)
+        old_cats = cudf.DataFrame._from_data(
+            {"cats": column.as_column(replaced.dtype.categories)}
+        )
 
         # Create a column with the appropriate labels replaced
-        old_cats["cats_replace"] = old_cats["cats"].replace(
+        old_cats._data["cats_replace"] = old_cats["cats"].replace(
```
[the remark this answers is not in the document]
Comment (galipremsagar) [reply python/cudf/cudf/core/column/categorical.py]: Yeah, changed these calls accordingly. Thanks for the catch @vyasr

File: python/cudf/cudf/core/column/categorical.py [hunk 14]
```diff
@@ -1095,18 +1094,21 @@ def find_and_replace(
...
-        new_cats = cudf.DataFrame(
-            {"cats": new_cats._data["cats"].apply_boolean_mask(bmask)}
-        ).reset_index()
+        bmask = old_cats._data["new_cats"].notnull()
+        new_cats = cudf.DataFrame._from_data(
+            {"cats": old_cats._data["new_cats"].apply_boolean_mask(bmask)}
+        )
+        new_cats._data["index"] = cudf.core.column.arange(
+            len(new_cats._data["cats"])
+        )
```
[the remark this answers is not in the document]
Comment (galipremsagar) [reply python/cudf/cudf/core/column/categorical.py]: Done

File: python/cudf/cudf/core/dataframe.py (hunk 7 above)
Comment (galipremsagar) [reply python/cudf/cudf/core/dataframe.py:1333]: We actually need to go through `__setitem__` calls for this user-facing API unfortunately 😞 . The reason is we want the scalar handling, index handling, etc. to match what pandas support.

File: python/cudf/cudf/core/index.py [hunk 15]
```diff
@@ -1,4 +1,4 @@
-# Copyright (c) 2018-2021, NVIDIA CORPORATION.
+# Copyright (c) 2018-2022, NVIDIA CORPORATION.
```
[the remark this answers is not in the document]
Comment (galipremsagar) [reply python/cudf/cudf/core/index.py]: Removed

File: python/cudf/cudf/core/multiindex.py (hunk 12 above)
Comment (galipremsagar) [reply python/cudf/cudf/core/multiindex.py]: Removed

File: python/cudf/cudf/core/column/categorical.py [hunk 16]
```diff
@@ -1507,9 +1509,15 @@ def _set_categories(
...
         new_codes = column.arange(len(new_cats), dtype=out_code_dtype)
 
-        new_df = cudf.DataFrame({"new_codes": new_codes, "cats": new_cats})
-        old_df = cudf.DataFrame({"old_codes": old_codes, "cats": cur_cats})
-        cur_df = cudf.DataFrame({"old_codes": cur_codes, "order": cur_order})
+        new_df = cudf.DataFrame._from_data(
+            data={"new_codes": new_codes, "cats": new_cats}
+        )
+        old_df = cudf.DataFrame._from_data(
+            data={"old_codes": old_codes, "cats": cur_cats.copy()}
```
[the remark this answers is not in the document]
Comment (galipremsagar) [reply python/cudf/cudf/core/column/categorical.py]: yeah not needed actually.

**Review by bdice — APPROVED** [review]
This is a nice improvement!

File: python/cudf/cudf/core/column/categorical.py [hunk 17]
```diff
@@ -1078,35 +1079,41 @@ def find_and_replace(
...
+                    to_replace_col, replacement_col
+                ),
+            }
         )
 
         # Construct the new categorical labels
         # If a category is being replaced by an existing one, we
         # want to map it to None. If it's totally new, we want to
         # map it to the new label it is to be replaced by
         dtype_replace = cudf.Series(replacement_col)
```
Comment (vyasr) [inline python/cudf/cudf/core/column/categorical.py:1099]: This is actually probably another opportunity to use _from_data.

File: python/cudf/cudf/core/column/numerical.py [hunk 18]
```diff
@@ -350,12 +376,21 @@ def _can_return_nan(self, skipna: bool = None) -> bool:
...
+            "cudf.core.column.ColumnBase",
+            (
+                super(NumericalColumn, self.nans_to_nulls())
+                if skipna
+                else super()
+            ),
+        )
+        return super_obj._process_for_reduction(
             skipna=skipna, min_count=min_count
         )
```
Comment (vyasr) [inline python/cudf/cudf/core/column/numerical.py:395]: I chose not to suggest this change last time around, but since @bdice started it :P
```suggestion
        col = self.nans_to_nulls() if skipna else self
        return super(NumericalColumn, col)._process_for_reduction(
            skipna=skipna, min_count=min_count
        )
```
This saves you some of the ugly mypy casting and having to save the output of super calls.

File: python/cudf/cudf/core/column/categorical.py (hunk 17 above)
Comment (galipremsagar) [reply python/cudf/cudf/core/column/categorical.py]: Sure, done

**Comment by galipremsagar** [comment]
@gpucibot merge
